ClusterInfoTest.all_nodes_should_have_the_same_cluster_start_time_and_cluster_id

```
java.lang.AssertionError: expected:<1462416052225> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.cluster.ClusterInfoTest.all_nodes_should_have_the_same_cluster_start_time_and_cluster_id(ClusterInfoTest.java:72)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-OracleJDK1.6/com.hazelcast$hazelcast/899/testReport/junit/com.hazelcast.cluster/ClusterInfoTest/all_nodes_should_have_the_same_cluster_start_time_and_cluster_id/

```
java.lang.AssertionError: expected:<1462979987682> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.cluster.ClusterInfoTest.all_nodes_should_have_the_same_cluster_start_time_and_cluster_id(ClusterInfoTest.java:71)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x/com.hazelcast$hazelcast/5064/testReport/junit/com.hazelcast.cluster/ClusterInfoTest/all_nodes_should_have_the_same_cluster_start_time_and_cluster_id/

--- a/hazelcast/src/main/java/com/hazelcast/cluster/impl/TcpIpJoiner.java
+++ b/hazelcast/src/main/java/com/hazelcast/cluster/impl/TcpIpJoiner.java
@@ -43,6 +43,7 @@ import java.util.LinkedList;
 import java.util.Set;
 import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
 
 import static com.hazelcast.util.AddressUtil.AddressHolder;
 
@@ -98,8 +99,8 @@ public class TcpIpJoiner extends AbstractJoiner {
             if (targetAddress == null) {
                 throw new IllegalArgumentException("Invalid target address -> NULL");
             }
-            if (logger.isFinestEnabled()) {
-                logger.finest("Joining over target member " + targetAddress);
+            if (logger.isFineEnabled()) {
+                logger.fine("Joining over target member " + targetAddress);
             }
             if (targetAddress.equals(node.getThisAddress()) || isLocalAddress(targetAddress)) {
                 node.setAsMaster();
@@ -116,8 +117,8 @@ public class TcpIpJoiner extends AbstractJoiner {
                     Thread.sleep(JOIN_RETRY_WAIT_TIME);
                     continue;
                 }
-                if (logger.isFinestEnabled()) {
-                    logger.finest("Sending joinRequest " + targetAddress);
+                if (logger.isFineEnabled()) {
+                    logger.fine("Sending joinRequest " + targetAddress);
                 }
                 clusterJoinManager.sendJoinRequest(targetAddress, true);
                 //noinspection BusyWait
@@ -135,7 +136,7 @@ public class TcpIpJoiner extends AbstractJoiner {
 
             boolean foundConnection = tryInitialConnection(possibleAddresses);
             if (!foundConnection) {
-                logger.finest("This node will assume master role since no possible member where connected to.");
+                logger.fine("This node will assume master role since no possible member where connected to.");
                 node.setAsMaster();
                 return;
             }
@@ -152,7 +153,7 @@ public class TcpIpJoiner extends AbstractJoiner {
                 }
 
                 if (isAllBlacklisted(possibleAddresses)) {
-                    logger.finest(
+                    logger.fine(
                             "This node will assume master role since none of the possible members accepted join request.");
                     node.setAsMaster();
                     return;
@@ -162,10 +163,10 @@ public class TcpIpJoiner extends AbstractJoiner {
                 if (masterCandidate) {
                     boolean consensus = claimMastership(possibleAddresses);
                     if (consensus) {
-                        if (logger.isFinestEnabled()) {
+                        if (logger.isFineEnabled()) {
                             Set<Address> votingEndpoints = new HashSet<Address>(possibleAddresses);
                             votingEndpoints.removeAll(blacklistedAddresses.keySet());
-                            logger.finest("Setting myself as master after consensus!"
+                            logger.fine("Setting myself as master after consensus!"
                                     + " Voting endpoints: " + votingEndpoints);
                         }
                         node.setAsMaster();
@@ -173,8 +174,8 @@ public class TcpIpJoiner extends AbstractJoiner {
                         return;
                     }
                 } else {
-                    if (logger.isFinestEnabled()) {
-                        logger.finest("Cannot claim myself as master! Will try to connect a possible master...");
+                    if (logger.isFineEnabled()) {
+                        logger.fine("Cannot claim myself as master! Will try to connect a possible master...");
                     }
                 }
 
@@ -188,10 +189,10 @@ public class TcpIpJoiner extends AbstractJoiner {
 
     @SuppressWarnings("checkstyle:npathcomplexity")
     private boolean claimMastership(Collection<Address> possibleAddresses) {
-        if (logger.isFinestEnabled()) {
+        if (logger.isFineEnabled()) {
             Set<Address> votingEndpoints = new HashSet<Address>(possibleAddresses);
             votingEndpoints.removeAll(blacklistedAddresses.keySet());
-            logger.finest("Claiming myself as master node! Asking to endpoints: " + votingEndpoints);
+            logger.fine("Claiming myself as master node! Asking to endpoints: " + votingEndpoints);
         }
         claimingMaster = true;
         Collection<Future<Boolean>> responses = new LinkedList<Future<Boolean>>();
@@ -259,8 +260,8 @@ public class TcpIpJoiner extends AbstractJoiner {
             }
 
             if (masterAddress != null) {
-                if (logger.isFinestEnabled()) {
-                    logger.finest("Sending join request to " + masterAddress);
+                if (logger.isFineEnabled()) {
+                    logger.fine("Sending join request to " + masterAddress);
                 }
                 clusterJoinManager.sendJoinRequest(masterAddress, true);
             } else {
@@ -280,8 +281,8 @@ public class TcpIpJoiner extends AbstractJoiner {
             if (isAllBlacklisted(possibleAddresses)) {
                 return false;
             }
-            if (logger.isFinestEnabled()) {
-                logger.finest("Will send master question to each address in: " + possibleAddresses);
+            if (logger.isFineEnabled()) {
+                logger.fine("Will send master question to each address in: " + possibleAddresses);
             }
             if (sendMasterQuestion(possibleAddresses)) {
                 return true;
@@ -312,8 +313,8 @@ public class TcpIpJoiner extends AbstractJoiner {
         }
 
         if (isAllBlacklisted(possibleAddresses) && node.getMasterAddress() == null) {
-            if (logger.isFinestEnabled()) {
-                logger.finest("Setting myself as master! No possible addresses remaining to connect...");
+            if (logger.isFineEnabled()) {
+                logger.fine("Setting myself as master! No possible addresses remaining to connect...");
             }
             node.setAsMaster();
             return;
@@ -326,8 +327,8 @@ public class TcpIpJoiner extends AbstractJoiner {
 
             Address master = node.getMasterAddress();
             if (master != null) {
-                if (logger.isFinestEnabled()) {
-                    logger.finest("Joining to master " + master);
+                if (logger.isFineEnabled()) {
+                    logger.fine("Joining to master " + master);
                 }
                 clusterJoinManager.sendJoinRequest(master, true);
             } else {
@@ -343,24 +344,24 @@ public class TcpIpJoiner extends AbstractJoiner {
             if (master != null) {
                 logger.warning("Couldn't join to the master : " + master);
             } else {
-                if (logger.isFinestEnabled()) {
-                    logger.finest("Couldn't find a master! But there was connections available: " + possibleAddresses);
+                if (logger.isFineEnabled()) {
+                    logger.fine("Couldn't find a master! But there was connections available: " + possibleAddresses);
                 }
             }
         }
     }
 
     private boolean sendMasterQuestion(Collection<Address> possibleAddresses) {
-        if (logger.isFinestEnabled()) {
-            logger.finest("NOT sending master question to blacklisted endpoints: " + blacklistedAddresses);
+        if (logger.isFineEnabled()) {
+            logger.fine("NOT sending master question to blacklisted endpoints: " + blacklistedAddresses);
         }
         boolean sent = false;
         for (Address address : possibleAddresses) {
             if (isBlacklisted(address)) {
                 continue;
             }
-            if (logger.isFinestEnabled()) {
-                logger.finest("Sending master question to " + address);
+            if (logger.isFineEnabled()) {
+                logger.fine("Sending master question to " + address);
             }
             if (clusterJoinManager.sendMasterQuestion(address)) {
                 sent = true;
@@ -444,8 +445,8 @@ public class TcpIpJoiner extends AbstractJoiner {
             } catch (UnknownHostException e) {
                 logger.warning("Cannot resolve hostname '" + addressHolder.getAddress()
                         + "'. Please make sure host is valid and reachable.");
-                if (logger.isFinestEnabled()) {
-                    logger.finest("Error during resolving possible target!", e);
+                if (logger.isFineEnabled()) {
+                    logger.log(Level.FINE, "Error during resolving possible target!", e);
                 }
             }
         }
@@ -477,8 +478,8 @@ public class TcpIpJoiner extends AbstractJoiner {
     private boolean isLocalAddress(final Address address) throws UnknownHostException {
         final Address thisAddress = node.getThisAddress();
         final boolean local = thisAddress.getInetSocketAddress().equals(address.getInetSocketAddress());
-        if (logger.isFinestEnabled()) {
-            logger.finest(address + " is local? " + local);
+        if (logger.isFineEnabled()) {
+            logger.fine(address + " is local? " + local);
         }
         return local;
     }

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/AbstractJoiner.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/AbstractJoiner.java
@@ -47,6 +47,7 @@ import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.logging.Level;
 
 import static com.hazelcast.cluster.memberselector.MemberSelectors.DATA_MEMBER_SELECTOR;
 import static com.hazelcast.spi.impl.OperationResponseHandlerFactory.createEmptyResponseHandler;
@@ -138,8 +139,8 @@ public abstract class AbstractJoiner implements Joiner {
     private void postJoin() {
         blacklistedAddresses.clear();
 
-        if (logger.isFinestEnabled()) {
-            logger.finest("PostJoin master: " + node.getMasterAddress() + ", isMaster: " + node.isMaster());
+        if (logger.isFineEnabled()) {
+            logger.fine("PostJoin master: " + node.getMasterAddress() + ", isMaster: " + node.isMaster());
         }
         if (node.getState() != NodeState.ACTIVE) {
             return;
@@ -163,7 +164,7 @@ public abstract class AbstractJoiner implements Joiner {
     private void ensureConnectionToAllMembers() {
         boolean allConnected = false;
         if (node.joined()) {
-            logger.finest("Waiting for all connections");
+            logger.fine("Waiting for all connections");
             int connectAllWaitSeconds = node.getProperties().getSeconds(GroupProperty.CONNECT_ALL_WAIT_SECONDS);
             int checkCount = 0;
             while (checkCount++ < connectAllWaitSeconds && !allConnected) {
@@ -179,8 +180,8 @@ public abstract class AbstractJoiner implements Joiner {
                 for (Member member : members) {
                     if (!member.localMember() && node.connectionManager.getOrConnect(member.getAddress()) == null) {
                         allConnected = false;
-                        if (logger.isFinestEnabled()) {
-                            logger.finest("Not-connected to " + member.getAddress());
+                        if (logger.isFineEnabled()) {
+                            logger.fine("Not-connected to " + member.getAddress());
                         }
                     }
                 }
@@ -210,19 +211,19 @@ public abstract class AbstractJoiner implements Joiner {
         try {
             boolean validJoinRequest = clusterJoinManager.validateJoinMessage(joinMessage);
             if (!validJoinRequest) {
-                logger.finest("Cannot process split brain merge message from " + joinMessage.getAddress()
+                logger.fine("Cannot process split brain merge message from " + joinMessage.getAddress()
                         + ", since join-message could not be validated.");
                 return false;
             }
         } catch (Exception e) {
-            logger.finest(e.getMessage());
+            logger.log(Level.FINE, "failure during validating join message", e);
             return false;
         }
 
         try {
             if (clusterService.getMember(joinMessage.getAddress()) != null) {
-                if (logger.isFinestEnabled()) {
-                    logger.finest("Should not merge to " + joinMessage.getAddress()
+                if (logger.isFineEnabled()) {
+                    logger.fine("Should not merge to " + joinMessage.getAddress()
                             + ", because it is already member of this cluster.");
                 }
                 return false;
@@ -230,8 +231,8 @@ public abstract class AbstractJoiner implements Joiner {
 
             final ClusterState clusterState = clusterService.getClusterState();
             if (clusterState != ClusterState.ACTIVE) {
-                if (logger.isFinestEnabled()) {
-                    logger.finest("Should not merge to " + joinMessage.getAddress() + ", because this cluster is in "
+                if (logger.isFineEnabled()) {
+                    logger.fine("Should not merge to " + joinMessage.getAddress() + ", because this cluster is in "
                             + clusterState + " state.");
                 }
                 return false;
@@ -264,8 +265,8 @@ public abstract class AbstractJoiner implements Joiner {
                 logger.info(node.getThisAddress() + " is merging to " + joinMessage.getAddress()
                         + ", because : joinMessage.getDataMemberCount() > currentDataMemberCount ["
                         + (targetDataMemberCount + " > " + currentDataMemberCount) + ']');
-                if (logger.isFinestEnabled()) {
-                    logger.finest(joinMessage.toString());
+                if (logger.isFineEnabled()) {
+                    logger.fine(joinMessage.toString());
                 }
                 return true;
             } else if (targetDataMemberCount == currentDataMemberCount) {
@@ -274,8 +275,8 @@ public abstract class AbstractJoiner implements Joiner {
                     logger.info(node.getThisAddress() + " is merging to " + joinMessage.getAddress()
                             + ", because : node.getThisAddress().hashCode() > joinMessage.address.hashCode() "
                             + ", this node member count: " + currentDataMemberCount);
-                    if (logger.isFinestEnabled()) {
-                        logger.finest(joinMessage.toString());
+                    if (logger.isFineEnabled()) {
+                        logger.fine(joinMessage.toString());
                     }
                     return true;
                 } else {
@@ -295,8 +296,8 @@ public abstract class AbstractJoiner implements Joiner {
     }
 
     protected JoinMessage sendSplitBrainJoinMessage(Address target) {
-        if (logger.isFinestEnabled()) {
-            logger.finest(node.getThisAddress() + " is connecting to " + target);
+        if (logger.isFineEnabled()) {
+            logger.fine(node.getThisAddress() + " is connecting to " + target);
         }
 
         Connection conn = node.connectionManager.getOrConnect(target, true);
@@ -323,7 +324,7 @@ public abstract class AbstractJoiner implements Joiner {
         try {
             return (JoinMessage) future.get(SPLIT_BRAIN_JOIN_CHECK_TIMEOUT_SECONDS, TimeUnit.SECONDS);
         } catch (TimeoutException e) {
-            logger.finest("Timeout during join check!", e);
+            logger.log(Level.FINE, "Timeout during join check!", e);
         } catch (Exception e) {
             logger.warning("Error during join check!", e);
         }

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterClockImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterClockImpl.java
@@ -50,8 +50,8 @@ public class ClusterClockImpl implements ClusterClock {
     }
 
     void setClusterTimeDiff(long diff) {
-        if (logger.isFinestEnabled()) {
-            logger.finest("Setting cluster time diff to " + diff + "ms.");
+        if (logger.isFineEnabled()) {
+            logger.fine("Setting cluster time diff to " + diff + "ms.");
         }
 
         if (abs(diff) > abs(maxClusterTimeDiff)) {

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterHeartbeatManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterHeartbeatManager.java
@@ -160,8 +160,8 @@ public class ClusterHeartbeatManager {
 
     public void acceptMasterConfirmation(MemberImpl member, long timestamp) {
         if (member != null) {
-            if (logger.isFinestEnabled()) {
-                logger.finest("MasterConfirmation has been received from " + member);
+            if (logger.isFineEnabled()) {
+                logger.fine("MasterConfirmation has been received from " + member);
             }
             long clusterTime = clusterClock.getClusterTime();
             if (clusterTime - timestamp > maxNoMasterConfirmationMillis / 2) {
@@ -266,8 +266,8 @@ public class ClusterHeartbeatManager {
             clusterService.removeAddress(member.getAddress(), reason);
             return true;
         }
-        if (logger.isFinestEnabled() && (now - heartbeatTime) > heartbeatIntervalMillis * HEART_BEAT_INTERVAL_FACTOR) {
-            logger.finest(format("Not receiving any heartbeats from %s since %s", member, timeToString(heartbeatTime)));
+        if (logger.isFineEnabled() && (now - heartbeatTime) > heartbeatIntervalMillis * HEART_BEAT_INTERVAL_FACTOR) {
+            logger.fine(format("Not receiving any heartbeats from %s since %s", member, timeToString(heartbeatTime)));
         }
         return false;
     }
@@ -371,8 +371,8 @@ public class ClusterHeartbeatManager {
         try {
             node.nodeEngine.getOperationService().send(new HeartbeatOperation(clusterClock.getClusterTime()), target);
         } catch (Exception e) {
-            if (logger.isFinestEnabled()) {
-                logger.finest(format("Error while sending heartbeat -> %s[%s]", e.getClass().getName(), e.getMessage()));
+            if (logger.isFineEnabled()) {
+                logger.fine(format("Error while sending heartbeat -> %s[%s]", e.getClass().getName(), e.getMessage()));
             }
         }
     }
@@ -398,16 +398,16 @@ public class ClusterHeartbeatManager {
         }
         Address masterAddress = node.getMasterAddress();
         if (masterAddress == null) {
-            logger.finest("Could not send MasterConfirmation, masterAddress is null!");
+            logger.fine("Could not send MasterConfirmation, masterAddress is null!");
             return;
         }
         MemberImpl masterMember = clusterService.getMember(masterAddress);
         if (masterMember == null) {
-            logger.finest("Could not send MasterConfirmation, masterMember is null!");
+            logger.fine("Could not send MasterConfirmation, masterMember is null!");
             return;
         }
-        if (logger.isFinestEnabled()) {
-            logger.finest("Sending MasterConfirmation to " + masterMember);
+        if (logger.isFineEnabled()) {
+            logger.fine("Sending MasterConfirmation to " + masterMember);
         }
         nodeEngine.getOperationService().send(new MasterConfirmationOperation(clusterClock.getClusterTime()),
                 masterAddress);

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java
@@ -454,8 +454,8 @@ public class ClusterJoinManager {
                 PartitionRuntimeState partitionRuntimeState = node.getPartitionService().createPartitionState();
 
                 Operation operation = new FinalizeJoinOperation(createMemberInfoList(clusterService.getMemberImpls()),
-                        postJoinOp, clusterClock.getClusterTime(), clusterStateManager.getState(),
-                        partitionRuntimeState, false);
+                        postJoinOp, clusterClock.getClusterTime(), clusterService.getClusterId(),
+                        clusterClock.getClusterStartTime(), clusterStateManager.getState(), partitionRuntimeState, false);
                 nodeEngine.getOperationService().send(operation, target);
             } else {
                 sendMasterAnswer(target);

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java
@@ -520,10 +520,11 @@ public class ClusterJoinManager {
                     calls.add(invokeClusterOperation(joinOperation, member.getAddress()));
                 }
                 for (MemberImpl member : members) {
-                    if (!member.getAddress().equals(clusterService.getThisAddress())) {
-                        Operation infoUpdateOperation = new MemberInfoUpdateOperation(memberInfos, time, true);
-                        calls.add(invokeClusterOperation(infoUpdateOperation, member.getAddress()));
+                    if (member.localMember() || joiningMembers.containsKey(member.getAddress())) {
+                        continue;
                     }
+                    Operation infoUpdateOperation = new MemberInfoUpdateOperation(memberInfos, time, true);
+                    calls.add(invokeClusterOperation(infoUpdateOperation, member.getAddress()));
                 }
 
                 int timeout = Math.min(calls.size() * FINALIZE_JOIN_TIMEOUT_FACTOR, FINALIZE_JOIN_MAX_TIMEOUT);

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java
@@ -141,8 +141,8 @@ public class ClusterJoinManager {
         }
 
         if (joinInProgress) {
-            if (logger.isFinestEnabled()) {
-                logger.finest(format("Join is in progress, cannot handle join request from %s at the moment", target));
+            if (logger.isFineEnabled()) {
+                logger.fine(format("Join is in progress, cannot handle join request from %s at the moment", target));
             }
             return;
         }
@@ -154,8 +154,8 @@ public class ClusterJoinManager {
         if (node.joined() && node.isRunning()) {
             return true;
         }
-        if (logger.isFinestEnabled()) {
-            logger.finest("Node is not ready to process join request...");
+        if (logger.isFineEnabled()) {
+            logger.fine("Node is not ready to process join request...");
         }
         return false;
     }
@@ -299,9 +299,9 @@ public class ClusterJoinManager {
 
     private void startJoinRequest(MemberInfo memberInfo) {
         long now = Clock.currentTimeMillis();
-        if (logger.isFinestEnabled()) {
+        if (logger.isFineEnabled()) {
             String timeToStart = (timeToStartJoin > 0 ? ", timeToStart: " + (timeToStartJoin - now) : "");
-            logger.finest(format("Handling join from %s, joinInProgress: %b%s", memberInfo.getAddress(),
+            logger.fine(format("Handling join from %s, joinInProgress: %b%s", memberInfo.getAddress(),
                     joinInProgress, timeToStart));
         }
 
@@ -335,8 +335,8 @@ public class ClusterJoinManager {
 
     public void handleMaster(Address masterAddress, Address callerAddress) {
         if (node.joined()) {
-            if (logger.isFinestEnabled()) {
-                logger.finest(format("Ignoring master response %s from %s, this node is already joined",
+            if (logger.isFineEnabled()) {
+                logger.fine(format("Ignoring master response %s from %s, this node is already joined",
                         masterAddress, callerAddress));
             }
             return;
@@ -344,7 +344,7 @@ public class ClusterJoinManager {
 
         if (node.getThisAddress().equals(masterAddress)) {
             if (node.isMaster()) {
-                logger.finest(format("Ignoring master response %s from %s, this node is already master",
+                logger.fine(format("Ignoring master response %s from %s, this node is already master",
                         masterAddress, callerAddress));
             } else {
                 node.setAsMaster();
@@ -358,8 +358,8 @@ public class ClusterJoinManager {
     private void handleMasterResponse(Address masterAddress, Address callerAddress) {
         clusterServiceLock.lock();
         try {
-            if (logger.isFinestEnabled()) {
-                logger.finest(format("Handling master response %s from %s", masterAddress, callerAddress));
+            if (logger.isFineEnabled()) {
+                logger.fine(format("Handling master response %s from %s", masterAddress, callerAddress));
             }
 
             Address currentMaster = node.getMasterAddress();
@@ -417,8 +417,8 @@ public class ClusterJoinManager {
                 sendMasterAnswer(joinMessage.getAddress());
             }
         } else {
-            if (logger.isFinestEnabled()) {
-                logger.finest(format("Received a master question from %s,"
+            if (logger.isFineEnabled()) {
+                logger.fine(format("Received a master question from %s,"
                         + " but this node is not master itself or doesn't have a master yet!", joinMessage.getAddress()));
             }
         }
@@ -443,8 +443,8 @@ public class ClusterJoinManager {
         Address target = member.getAddress();
         if (joinMessage.getUuid().equals(member.getUuid())) {
             if (node.isMaster()) {
-                if (logger.isFinestEnabled()) {
-                    logger.finest(format("Ignoring join request, member already exists: %s", joinMessage));
+                if (logger.isFineEnabled()) {
+                    logger.fine(format("Ignoring join request, member already exists: %s", joinMessage));
                 }
 
                 // send members update back to node trying to join again...
@@ -486,7 +486,7 @@ public class ClusterJoinManager {
     }
 
     private void startJoin() {
-        logger.finest("Starting join...");
+        logger.fine("Starting join...");
         clusterServiceLock.lock();
         try {
             try {

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterServiceImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterServiceImpl.java
@@ -302,8 +302,8 @@ public class ClusterServiceImpl implements ClusterService, ConnectionListener, M
             node.setMasterAddress(null);
         }
 
-        if (logger.isFinestEnabled()) {
-            logger.finest(format("Old master: %s, new master: %s ", oldMasterAddress, node.getMasterAddress()));
+        if (logger.isFineEnabled()) {
+            logger.fine(format("Old master: %s, new master: %s ", oldMasterAddress, node.getMasterAddress()));
         }
 
         if (node.isMaster()) {
@@ -405,7 +405,7 @@ public class ClusterServiceImpl implements ClusterService, ConnectionListener, M
         if (currentMembersSize == newMembersSize) {
             Set<MemberInfo> currentMemberInfos = createMemberInfoSet(currentMembers.values());
             if (currentMemberInfos.containsAll(newMemberInfos)) {
-                logger.finest("Received a periodic member update, no need to process...");
+                logger.fine("Received a periodic member update, no need to process...");
             } else {
                 logger.warning("Received an inconsistent member update "
                         + "which contains new members and removes some of the current members! "
@@ -476,8 +476,8 @@ public class ClusterServiceImpl implements ClusterService, ConnectionListener, M
 
     @Override
     public void connectionRemoved(Connection connection) {
-        if (logger.isFinestEnabled()) {
-            logger.finest("Removed connection " + connection.getEndPoint());
+        if (logger.isFineEnabled()) {
+            logger.fine("Removed connection " + connection.getEndPoint());
         }
         if (!node.joined()) {
             Address masterAddress = node.getMasterAddress();
@@ -495,8 +495,8 @@ public class ClusterServiceImpl implements ClusterService, ConnectionListener, M
         if (members == null || members.length == 0) {
             return;
         }
-        if (logger.isFinestEnabled()) {
-            logger.finest("Updating members " + Arrays.toString(members));
+        if (logger.isFineEnabled()) {
+            logger.fine("Updating members " + Arrays.toString(members));
         }
         lock.lock();
         try {
@@ -526,16 +526,16 @@ public class ClusterServiceImpl implements ClusterService, ConnectionListener, M
                 setMembersRef(newMembers);
 
                 if (node.isMaster()) {
-                    if (logger.isFinestEnabled()) {
-                        logger.finest(deadMember + " is dead, sending remove to all other members...");
+                    if (logger.isFineEnabled()) {
+                        logger.fine(deadMember + " is dead, sending remove to all other members...");
                     }
                     sendMemberRemoveOperation(deadMember);
                 }
 
                 final ClusterState clusterState = clusterStateManager.getState();
                 if (clusterState != ClusterState.ACTIVE) {
-                    if (logger.isFinestEnabled()) {
-                        logger.finest(deadMember + " is dead, added to members left while cluster is " + clusterState + " state");
+                    if (logger.isFineEnabled()) {
+                        logger.fine(deadMember + " is dead, added to members left while cluster is " + clusterState + " state");
                     }
                     Map<Address, MemberImpl> membersRemovedInNotActiveState
                             = new LinkedHashMap<Address, MemberImpl>(membersRemovedInNotActiveStateRef.get());

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterServiceImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterServiceImpl.java
@@ -382,8 +382,8 @@ public class ClusterServiceImpl implements ClusterService, ConnectionListener, M
             setMembers(updatedMembers);
             sendMembershipEvents(currentMemberMap.values(), newMembers);
 
-            clusterHeartbeatManager.heartBeat();
             node.setJoined();
+            clusterHeartbeatManager.heartBeat();
             logger.info(membersString());
         } finally {
             lock.unlock();

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterStateManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterStateManager.java
@@ -43,6 +43,7 @@ import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.concurrent.locks.Lock;
+import java.util.logging.Level;
 
 import static com.hazelcast.internal.cluster.impl.ClusterServiceImpl.SERVICE_NAME;
 import static com.hazelcast.util.FutureUtil.waitWithDeadline;
@@ -362,8 +363,8 @@ public class ClusterStateManager {
         }
 
         private void log(Throwable cause) {
-            if (logger.isFinestEnabled()) {
-                logger.finest(cause);
+            if (logger.isFineEnabled()) {
+                logger.log(Level.FINE, "failure during cluster state change", cause);
             }
         }
 

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MulticastJoiner.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MulticastJoiner.java
@@ -82,8 +82,8 @@ public class MulticastJoiner extends AbstractJoiner {
 
             Address master = node.getMasterAddress();
             if (master != null) {
-                if (logger.isFinestEnabled()) {
-                    logger.finest("Joining to master " + master);
+                if (logger.isFineEnabled()) {
+                    logger.fine("Joining to master " + master);
                 }
                 clusterJoinManager.sendJoinRequest(master, true);
             } else {
@@ -122,8 +122,8 @@ public class MulticastJoiner extends AbstractJoiner {
             JoinMessage joinInfo = q.poll(3, TimeUnit.SECONDS);
             if (joinInfo != null) {
                 if (node.clusterService.getMember(joinInfo.getAddress()) != null) {
-                    if (logger.isFinestEnabled()) {
-                        logger.finest("Ignoring merge join response, since " + joinInfo.getAddress()
+                    if (logger.isFineEnabled()) {
+                        logger.fine("Ignoring merge join response, since " + joinInfo.getAddress()
                                 + " is already a member.");
                     }
                     return;
@@ -165,8 +165,8 @@ public class MulticastJoiner extends AbstractJoiner {
 
     private Address findMasterWithMulticast() {
         try {
-            if (logger.isFinestEnabled()) {
-                logger.finest("Searching for master node. Max tries: " + maxTryCount.get());
+            if (logger.isFineEnabled()) {
+                logger.fine("Searching for master node. Max tries: " + maxTryCount.get());
             }
             JoinRequest joinRequest = node.createJoinRequest(false);
             while (node.isRunning() && currentTryCount.incrementAndGet() <= maxTryCount.get()) {

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/NodeMulticastListener.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/NodeMulticastListener.java
@@ -59,8 +59,8 @@ public class NodeMulticastListener implements MulticastListener {
     }
 
     private void logDroppedMessage(Object msg) {
-        if (logger.isFinestEnabled()) {
-            logger.finest("Dropped: " + msg);
+        if (logger.isFineEnabled()) {
+            logger.fine("Dropped: " + msg);
         }
     }
 
@@ -118,8 +118,8 @@ public class NodeMulticastListener implements MulticastListener {
     }
 
     private void logJoinMessageDropped(String masterHost) {
-        if (logger.isFinestEnabled()) {
-            logger.finest(format(
+        if (logger.isFineEnabled()) {
+            logger.fine(format(
                     "JoinMessage from %s is dropped because its sender is not a trusted interface", masterHost));
         }
     }

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/FinalizeJoinOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/FinalizeJoinOperation.java
@@ -84,8 +84,8 @@ public class FinalizeJoinOperation extends MemberInfoUpdateOperation implements 
 
         if (nodeEngine.getNode().joined()) {
             ILogger logger = getLogger();
-            if (logger.isFinestEnabled()) {
-                logger.finest("Node is already joined... No need to finalize join...");
+            if (logger.isFineEnabled()) {
+                logger.fine("Node is already joined... No need to finalize join...");
             }
             return;
         }

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/FinalizeJoinOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/FinalizeJoinOperation.java
@@ -66,9 +66,12 @@ public class FinalizeJoinOperation extends MemberInfoUpdateOperation implements 
     }
 
     public FinalizeJoinOperation(Collection<MemberInfo> members, PostJoinOperation postJoinOp, long masterTime,
-                                 ClusterState clusterState, PartitionRuntimeState partitionRuntimeState, boolean sendResponse) {
+                                 String clusterId, long clusterStartTime, ClusterState clusterState,
+                                 PartitionRuntimeState partitionRuntimeState, boolean sendResponse) {
         super(members, masterTime, sendResponse);
         this.postJoinOp = postJoinOp;
+        this.clusterId = clusterId;
+        this.clusterStartTime = clusterStartTime;
         this.clusterState = clusterState;
         this.partitionRuntimeState = partitionRuntimeState;
     }

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/HeartbeatOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/HeartbeatOperation.java
@@ -44,8 +44,8 @@ public final class HeartbeatOperation extends AbstractClusterOperation
         MemberImpl member = service.getMember(getCallerAddress());
         if (member == null) {
             ILogger logger = getLogger();
-            if (logger.isFinestEnabled()) {
-                logger.finest("Heartbeat received from an unknown endpoint: " + getCallerAddress());
+            if (logger.isFineEnabled()) {
+                logger.fine("Heartbeat received from an unknown endpoint: " + getCallerAddress());
             }
             return;
         }

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/JoinCheckOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/JoinCheckOperation.java
@@ -63,12 +63,12 @@ public class JoinCheckOperation extends AbstractOperation implements JoinOperati
                 if (service.getClusterJoinManager().validateJoinMessage(request)) {
                     response = node.createSplitBrainJoinMessage();
                 }
-                if (logger.isFinestEnabled()) {
-                    logger.finest("Returning " + response + " to " + getCallerAddress());
+                if (logger.isFineEnabled()) {
+                    logger.fine("Returning " + response + " to " + getCallerAddress());
                 }
             } catch (Exception e) {
-                if (logger.isFinestEnabled()) {
-                    logger.finest("Could not validate split-brain join message! -> " + e.getMessage());
+                if (logger.isFineEnabled()) {
+                    logger.fine("Could not validate split-brain join message! -> " + e.getMessage());
                 }
             }
         }

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/MasterClaimOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/MasterClaimOperation.java
@@ -44,8 +44,8 @@ public class MasterClaimOperation extends AbstractOperation implements JoinOpera
             approvedAsMaster = false;
             logger.warning("This node requires MulticastJoin strategy!");
         }
-        if (logger.isFinestEnabled()) {
-            logger.finest("Sending '" + approvedAsMaster + "' for master claim of node: " + getCallerAddress());
+        if (logger.isFineEnabled()) {
+            logger.fine("Sending '" + approvedAsMaster + "' for master claim of node: " + getCallerAddress());
         }
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/MemberInfoUpdateOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/MemberInfoUpdateOperation.java
@@ -19,11 +19,13 @@ package com.hazelcast.internal.cluster.impl.operations;
 import com.hazelcast.internal.cluster.MemberInfo;
 import com.hazelcast.internal.cluster.impl.ClusterDataSerializerHook;
 import com.hazelcast.internal.cluster.impl.ClusterServiceImpl;
+import com.hazelcast.logging.ILogger;
 import com.hazelcast.nio.Address;
 import com.hazelcast.nio.Connection;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.spi.impl.NodeEngineImpl;
 import com.hazelcast.util.Clock;
 
 import java.io.IOException;
@@ -48,6 +50,18 @@ public class MemberInfoUpdateOperation extends AbstractClusterOperation implemen
 
     @Override
     public void run() throws Exception {
+        final ClusterServiceImpl clusterService = getService();
+        final NodeEngineImpl nodeEngine = clusterService.getNodeEngine();
+
+        if (!nodeEngine.getNode().joined()) {
+            ILogger logger = getLogger();
+            if (logger.isFineEnabled()) {
+                logger.fine("Ignoring member info update since not joined yet...");
+            }
+
+            return;
+        }
+
         processMemberUpdate();
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/PostJoinOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/PostJoinOperation.java
@@ -28,6 +28,7 @@ import com.hazelcast.spi.UrgentSystemOperation;
 
 import java.io.IOException;
 import java.util.Arrays;
+import java.util.logging.Level;
 
 import static com.hazelcast.util.Preconditions.checkNegative;
 import static com.hazelcast.util.Preconditions.checkNotNull;
@@ -65,8 +66,8 @@ public class PostJoinOperation extends AbstractOperation implements UrgentSystem
                             logger.warning("Error while running post-join operation: "
                                     + t.getClass().getSimpleName() + ": " + t.getMessage());
 
-                            if (logger.isFinestEnabled()) {
-                                logger.finest(t);
+                            if (logger.isFineEnabled()) {
+                                logger.log(Level.FINE, "Error while running post-join operation: ", t);
                             }
                         }
                     }

--- a/hazelcast/src/main/java/com/hazelcast/nio/tcp/TcpIpConnectionManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/nio/tcp/TcpIpConnectionManager.java
@@ -239,7 +239,7 @@ public class TcpIpConnectionManager implements ConnectionManager, PacketHandler 
     /**
      * Binding completes the connection and makes it available to be used with the ConnectionManager.
      */
-    private boolean bind(TcpIpConnection connection, Address remoteEndPoint, Address localEndpoint, boolean reply) {
+    private synchronized boolean bind(TcpIpConnection connection, Address remoteEndPoint, Address localEndpoint, boolean reply) {
         if (logger.isFinestEnabled()) {
             logger.finest("Binding " + connection + " to " + remoteEndPoint + ", reply is " + reply);
         }
@@ -264,47 +264,51 @@ public class TcpIpConnectionManager implements ConnectionManager, PacketHandler 
 
     @Override
     public synchronized boolean registerConnection(final Address remoteEndPoint, final Connection connection) {
-        if (remoteEndPoint.equals(ioService.getThisAddress())) {
-            return false;
-        }
-
-        if (!connection.isAlive()) {
-            if (logger.isFinestEnabled()) {
-                logger.finest(connection + " to " + remoteEndPoint + " is not registered since connection is not active.");
+        try {
+            if (remoteEndPoint.equals(ioService.getThisAddress())) {
+                return false;
             }
-            return false;
-        }
 
-        if (connection instanceof TcpIpConnection) {
-            TcpIpConnection tcpConnection = (TcpIpConnection) connection;
-            Address currentEndPoint = tcpConnection.getEndPoint();
-            if (currentEndPoint != null && !currentEndPoint.equals(remoteEndPoint)) {
-                throw new IllegalArgumentException(connection + " has already a different endpoint than: "
-                        + remoteEndPoint);
+            if (!connection.isAlive()) {
+                if (logger.isFinestEnabled()) {
+                    logger.finest(connection + " to " + remoteEndPoint + " is not registered since connection is not active.");
+                }
+                return false;
             }
-            tcpConnection.setEndPoint(remoteEndPoint);
 
-            if (!connection.isClient()) {
-                TcpIpConnectionMonitor connectionMonitor = getConnectionMonitor(remoteEndPoint, true);
-                tcpConnection.setMonitor(connectionMonitor);
-            }
-        }
-        connectionsMap.put(remoteEndPoint, connection);
-        connectionsInProgress.remove(remoteEndPoint);
-        ioService.getEventService().executeEventCallback(new StripedRunnable() {
-            @Override
-            public void run() {
-                for (ConnectionListener listener : connectionListeners) {
-                    listener.connectionAdded(connection);
+            if (connection instanceof TcpIpConnection) {
+                TcpIpConnection tcpConnection = (TcpIpConnection) connection;
+                Address currentEndPoint = tcpConnection.getEndPoint();
+                if (currentEndPoint != null && !currentEndPoint.equals(remoteEndPoint)) {
+                    throw new IllegalArgumentException(connection + " has already a different endpoint than: "
+                            + remoteEndPoint);
+                }
+                tcpConnection.setEndPoint(remoteEndPoint);
+
+                if (!connection.isClient()) {
+                    TcpIpConnectionMonitor connectionMonitor = getConnectionMonitor(remoteEndPoint, true);
+                    tcpConnection.setMonitor(connectionMonitor);
                 }
             }
+            connectionsMap.put(remoteEndPoint, connection);
 
-            @Override
-            public int getKey() {
-                return remoteEndPoint.hashCode();
-            }
-        });
-        return true;
+            ioService.getEventService().executeEventCallback(new StripedRunnable() {
+                @Override
+                public void run() {
+                    for (ConnectionListener listener : connectionListeners) {
+                        listener.connectionAdded(connection);
+                    }
+                }
+
+                @Override
+                public int getKey() {
+                    return remoteEndPoint.hashCode();
+                }
+            });
+            return true;
+        } finally {
+            connectionsInProgress.remove(remoteEndPoint);
+        }
     }
 
     private boolean checkAlreadyConnected(TcpIpConnection connection, Address remoteEndPoint) {
@@ -343,25 +347,32 @@ public class TcpIpConnectionManager implements ConnectionManager, PacketHandler 
         return wrapper;
     }
 
-    TcpIpConnection newConnection(SocketChannelWrapper channel, Address endpoint) {
-        TcpIpConnection connection = new TcpIpConnection(
-                this,
-                connectionIdGen.incrementAndGet(),
-                channel,
-                ioThreadingModel);
+    synchronized TcpIpConnection newConnection(SocketChannelWrapper channel, Address endpoint) {
+        try {
+            if (!live) {
+                throw new IllegalStateException("connection manager is not live!");
+            }
 
-        connection.setEndPoint(endpoint);
-        activeConnections.add(connection);
-        acceptedSockets.remove(channel);
+            TcpIpConnection connection = new TcpIpConnection(
+                    this,
+                    connectionIdGen.incrementAndGet(),
+                    channel,
+                    ioThreadingModel);
 
-        connection.start();
-        ioThreadingModel.onConnectionAdded(connection);
+            connection.setEndPoint(endpoint);
+            activeConnections.add(connection);
 
-        logger.info("Established socket connection between "
-                + channel.socket().getLocalSocketAddress() + " and " + channel.socket().getRemoteSocketAddress());
-        openedCount.inc();
+            connection.start();
+            ioThreadingModel.onConnectionAdded(connection);
 
-        return connection;
+            logger.info("Established socket connection between "
+                    + channel.socket().getLocalSocketAddress() + " and " + channel.socket().getRemoteSocketAddress());
+            openedCount.inc();
+
+            return connection;
+        } finally {
+            acceptedSockets.remove(channel);
+        }
     }
 
     void failedConnection(Address address, Throwable t, boolean silent) {

